File: JobApplicationMicroservice/src/main/java/com/licenta/jobapplicationmicroservice/business/service/DatabaseService.java
package com.licenta.jobapplicationmicroservice.business.service;

import com.google.gson.Gson;
import com.licenta.jobapplicationmicroservice.business.interfaces.IDatabaseService;
import com.licenta.jobapplicationmicroservice.business.model.Job;
import com.licenta.jobapplicationmicroservice.business.util.errorhandler.RestTemplateResponseErrorHandler;
import com.licenta.jobapplicationmicroservice.business.util.exception.ExceptionWithStatus;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

@Service
public class DatabaseService implements IDatabaseService {

    private final RestTemplate restTemplate;
    private final String URL_FORMAT = "http://localhost:23050/%s/%s";

    @Autowired
    public DatabaseService() {
        restTemplate = new RestTemplateBuilder()
                .errorHandler(new RestTemplateResponseErrorHandler())
                .build();
    }

    @Override
    public void verifyUserExists(Long userId) {
        ResponseEntity<String> response = restTemplate.getForEntity(String.format(URL_FORMAT, "users", userId), String.class);

        if(response.getStatusCode() != HttpStatus.OK)
            throw new ExceptionWithStatus(response.getBody(), response.getStatusCode());
    }

    @Override
    public Job getJob(Long jobId) {
        ResponseEntity<String> response = restTemplate.getForEntity(String.format(URL_FORMAT, "jobs", jobId), String.class);

        if(response.getStatusCode() != HttpStatus.OK)
            throw new ExceptionWithStatus(response.getBody(), response.getStatusCode());

        Gson gson = new Gson();
        return gson.fromJson(response.getBody(), Job.class);
    }
}

File: SearchMicroservice/src/main/kotlin/com/licenta/searchmicroservice/business/service/DatabaseService.kt
package com.licenta.searchmicroservice.business.service

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.licenta.searchmicroservice.business.`interface`.IDatabaseService
import com.licenta.searchmicroservice.business.model.Job
import com.licenta.searchmicroservice.business.util.exception.ExceptionWithStatus
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class DatabaseService: IDatabaseService{

    override fun getJobs(): List<Job> {
        // TODO: replace localhost with env variable
        val response = khttp.get("http://localhost:23050/jobs")

        if (response.statusCode != 200) {
            throw ExceptionWithStatus(response.text, HttpStatus.valueOf(response.statusCode))
        }

        val typeToken = object : TypeToken<List<Job>>() {}.type

        return Gson().fromJson(response.text, typeToken)
    }
}